### PR TITLE
Sycl Kernel Support for Shuffle Primitive

### PR DESCRIPTION
**Pull request by ManojkumarGawali**
# Description

Introducing SYCL backend for oneDNN Shuffle primitive. This contains support for Shuffle functionality.

# **Supported scope**
Supported Data Types: f32, bf16, f16

## General

- [x] Do all unit and benchdnn tests (`make test` and `make test_benchdnn_*`) pass locally for each commit?
- [x] Have you formatted the code using clang-format?

# **Testing Scope**
         benchdnn: Passed

Please refer the attached test output files.
[test_shuffle.zip](https://github.com/oneapi-src/oneDNN/files/10879962/test_shuffle.zip)

### CODE REVIEW

File: src/gpu/sycl/ref_shuffle.cpp [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+#include "gpu/sycl/shuffle_kernels.hpp"
+#include "gpu/sycl/sycl_primitive_conf.hpp"
+
+namespace dnnl {
+namespace impl {
+namespace gpu {
+namespace sycl {
+
+using namespace impl::sycl;
+using namespace std;
```
Comment (densamoilov) [inline src/gpu/sycl/ref_shuffle.cpp:26]: please remote `using namespace std;`

File: src/gpu/sycl/ref_shuffle.cpp [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+status_t ref_shuffle_t::pd_t::init_conf() {
+    conf_ = sycl_shuffle_conf_t();
+    const memory_desc_wrapper data_d(src_md(0));
+    conf_.axis = axis();
+    conf_.axis_size = axis_size();
+    conf_.ndims = ndims();
+    auto dim = desc()->dst_desc.dims;
+    conf_.ndims_d = desc()->dst_desc.ndims;
+    conf_.outer_size = utils::array_product(dim, conf_.axis);
+    conf_.inner_size = utils::array_product(dim + conf_.axis + 1, conf_.ndims_d - conf_.axis - 1);
```
Comment (densamoilov) [inline src/gpu/sycl/ref_shuffle.cpp:37]: (random spot) please apply clang-format

File: src/gpu/sycl/ref_shuffle.cpp [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+status_t ref_shuffle_t::execute(const exec_ctx_t &ctx)const{
+parallel_for(ctx, kernel_, [&](::sycl::handler &cgh) {
+    auto src_mem_arg = pd()->is_fwd() ? CTX_IN_SYCL_KERNEL_MEMORY(DNNL_ARG_SRC)
+                                      : CTX_IN_SYCL_KERNEL_MEMORY(DNNL_ARG_DIFF_DST);
+    
+    auto dst_mem_arg =  pd()->is_fwd() 
+             ? CTX_OUT_SYCL_KERNEL_MEMORY(DNNL_ARG_DST) 
+             : CTX_OUT_SYCL_KERNEL_MEMORY(DNNL_ARG_DIFF_SRC); 
+    int axis = pd()->axis();
+    auto nelems_A = memory_desc_wrapper(pd()->src_md()).nelems();
```
Comment (densamoilov) [inline src/gpu/sycl/ref_shuffle.cpp:107]: There are a lot of unused variable warnings, please address them

File: src/gpu/sycl/ref_shuffle.hpp [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+        static bool check_formats(const memory_desc_wrapper &data) {
+            using namespace format_tag;
+            for (const auto &mdw : {data}) {
+                if (mdw.matches_one_of_tag(ab, abc, abcd, abcde) == undef) {
+                    return false;
+                }
+            }
+            return true;
+        }
+    };
```
Comment (densamoilov) [inline src/gpu/sycl/ref_shuffle.hpp:102]: These function are not used

File: src/gpu/sycl/ref_shuffle.hpp [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+
+        DECLARE_COMMON_PD_T("dpcpp:ref:any", ref_shuffle_t);
+
+        status_t init(engine_t *engine) {
+            using namespace data_type;
+            const memory_desc_wrapper data_d(src_md(0));
+            const memory_desc_wrapper dst_d(dst_md(0));
+
+            const data_type_t data_type = src_md()->data_type;
+            bool ok = is_fwd()
```
Comment (densamoilov) [inline src/gpu/sycl/ref_shuffle.hpp:53]: What is the reason for implementing only forward?

File: src/gpu/sycl/ref_shuffle.hpp [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+#include "sycl/sycl_stream.hpp"
+
+namespace dnnl {
+namespace impl {
+namespace gpu {
+namespace sycl {
+
+using namespace std;
+using namespace format_tag;
+using namespace utils;
```
Comment (densamoilov) [inline src/gpu/sycl/ref_shuffle.hpp:38]: Please don't use it in header files. Move them to the scope where they are used.

File: src/gpu/sycl/ref_shuffle.hpp [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+        status_t init(engine_t *engine) {
+            using namespace data_type;
+            const memory_desc_wrapper data_d(src_md(0));
+            const memory_desc_wrapper dst_d(dst_md(0));
+
+            const data_type_t data_type = src_md()->data_type;
+            bool ok = is_fwd()
+                    && (src_md()->data_type == bf16
+                            || src_md()->data_type == f32
+                            || src_md()->data_type == f16)
```
Comment (densamoilov) [inline src/gpu/sycl/ref_shuffle.hpp:56]: Shuffle also supports integer data types, is there any reason they are not supported?

File: src/gpu/sycl/shuffle_kernels.hpp [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+
+    void *data_ptr() const { return data_.get_pointer(); }
+    void *dst_ptr() const { return dst_.get_pointer(); }
+
+    sycl_shuffle_conf_t conf_;
+    sycl_in_memory_arg_t data_;
+    sycl_out_memory_arg_t dst_;
+    int transpose_col_;
+    int transpose_row_;
+    dnnl_dim_t groupsize_;
```
Comment (densamoilov) [inline src/gpu/sycl/shuffle_kernels.hpp:214]: these are not used

File: src/gpu/sycl/ref_shuffle.cpp [hunk 9]
```diff
@@ -0,0 +1,147 @@
+/*******************************************************************************
+* Copyright 2022 Intel Corporation
```
[the remark this answers is not in the document]
Comment (ManojkumarGawali) [reply src/gpu/sycl/ref_shuffle.cpp]: Updated

File: src/gpu/sycl/ref_shuffle.cpp (hunk 1 above)
Comment (ManojkumarGawali) [reply src/gpu/sycl/ref_shuffle.cpp]: Done

File: src/gpu/sycl/ref_shuffle.cpp (hunk 2 above)
Comment (ManojkumarGawali) [reply src/gpu/sycl/ref_shuffle.cpp]: Done

File: src/gpu/sycl/ref_shuffle.cpp (hunk 3 above)
Comment (ManojkumarGawali) [reply src/gpu/sycl/ref_shuffle.cpp]: Removed

File: src/gpu/sycl/ref_shuffle.hpp (hunk 4 above)
Comment (ManojkumarGawali) [reply src/gpu/sycl/ref_shuffle.hpp]: Removed

File: src/gpu/sycl/ref_shuffle.hpp (hunk 5 above)
Comment (ManojkumarGawali) [reply src/gpu/sycl/ref_shuffle.hpp]: Checked and updated

File: src/gpu/sycl/ref_shuffle.hpp (hunk 6 above)
Comment (ManojkumarGawali) [reply src/gpu/sycl/ref_shuffle.hpp]: Done

File: src/gpu/sycl/ref_shuffle.hpp (hunk 7 above)
Comment (ManojkumarGawali) [reply src/gpu/sycl/ref_shuffle.hpp]: Added int support

File: src/gpu/sycl/shuffle_kernels.hpp (hunk 8 above)
Comment (ManojkumarGawali) [reply src/gpu/sycl/shuffle_kernels.hpp]: Removed

**Pull request merged**